pdns/tcpiohandler.cc

```diff
@@ -0,0 +1,1269 @@
+#if defined(SSL_CTX_set_ecdh_auto)
+    SSL_CTX_set_ecdh_auto(d_tlsCtx.get(), 1);
+#endif
+
+    if (!params.d_ciphers.empty()) {
+      if (SSL_CTX_set_cipher_list(d_tlsCtx.get(), params.d_ciphers.c_str()) != 1) {
+        ERR_print_errors_fp(stderr);
+        throw std::runtime_error("Error setting the cipher list to '" + params.d_ciphers + "' for the TLS context");
+      }
+    }
+#ifdef HAVE_SSL_CTX_SET_CIPHERSUITES
+    if (!params.d_ciphers13.empty()) {
+      if (SSL_CTX_set_ciphersuites(d_tlsCtx.get(), params.d_ciphers13.c_str()) != 1) {
+        ERR_print_errors_fp(stderr);
+        throw std::runtime_error("Error setting the TLS 1.3 cipher list to '" + params.d_ciphers13 + "' for the TLS context");
+      }
+    }
+#endif /* HAVE_SSL_CTX_SET_CIPHERSUITES */
+
+    if (params.d_validateCertificates) {
+      if (params.d_caStore.empty())  {
+        if (SSL_CTX_set_default_verify_paths(d_tlsCtx.get()) != 1) {
+          throw std::runtime_error("Error adding the system's default trusted CAs");
+        }
+      } else {
+        if (SSL_CTX_load_verify_locations(d_tlsCtx.get(), params.d_caStore.c_str(), nullptr) != 1) {
+          throw std::runtime_error("Error adding the trusted CAs file " + params.d_caStore);
+        }
+      }
+
+      SSL_CTX_set_verify(d_tlsCtx.get(), SSL_VERIFY_PEER, nullptr);
+#if (OPENSSL_VERSION_NUMBER < 0x10002000L)
+      warnlog("TLS hostname validation requested but not supported for OpenSSL < 1.0.2");
+#endif
+    }
+  }
+
+  ~OpenSSLTLSIOCtx() override
+  {
+    d_tlsCtx.reset();
+    unregisterOpenSSLUser();
+  }
+
+  static int ticketKeyCb(SSL *s, unsigned char keyName[TLS_TICKETS_KEY_NAME_SIZE], unsigned char *iv, EVP_CIPHER_CTX *ectx, HMAC_CTX *hctx, int enc)
+  {
+    OpenSSLFrontendContext* ctx = reinterpret_cast<OpenSSLFrontendContext*>(libssl_get_ticket_key_callback_data(s));
+    if (ctx == nullptr) {
+      return -1;
+    }
+
+    int ret = libssl_ticket_key_callback(s, ctx->d_ticketKeys, keyName, iv, ectx, hctx, enc);
+    if (enc == 0) {
+      if (ret == 0 || ret == 2) {
+        OpenSSLTLSConnection* conn = reinterpret_cast<OpenSSLTLSConnection*>(SSL_get_ex_data(s, OpenSSLTLSConnection::s_tlsConnIndex));
+        if (conn) {
+          if (ret == 0) {
+            conn->setUnknownTicketKey();
+          }
+          else if (ret == 2) {
+            conn->setResumedFromInactiveTicketKey();
+          }
+        }
+      }
+    }
+
+    return ret;
+  }
+
+  static int ocspStaplingCb(SSL* ssl, void* arg)
+  {
+    if (ssl == nullptr || arg == nullptr) {
+      return SSL_TLSEXT_ERR_NOACK;
+    }
+    const auto ocspMap = reinterpret_cast<std::map<int, std::string>*>(arg);
+    return libssl_ocsp_stapling_callback(ssl, *ocspMap);
+  }
+
+  std::unique_ptr<TLSConnection> getConnection(int socket, unsigned int timeout, time_t now) override
+  {
+    handleTicketsKeyRotation(now);
+
+    return std::unique_ptr<OpenSSLTLSConnection>(new OpenSSLTLSConnection(socket, timeout, d_feContext));
```

Comment:
We have C++17 now :)
```suggestion
    return std::make_unique_ptr<OpenSSLTLSConnection>(socket, timeout, d_feContext);
```